slice/internal/controller/workload_controller.go

```diff
@@ -160,32 +159,143 @@ func (r *WorkloadReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 	return ctrl.Result{}, client.IgnoreNotFound(err)
 }
 
-func (r *WorkloadReconciler) shouldFinalize(wl *kueue.Workload) bool {
-	return !wl.DeletionTimestamp.IsZero() || workload.IsFinished(wl) || workload.IsEvicted(wl) || !workload.IsActive(wl)
+func shouldFinalize(wl *kueue.Workload) bool {
+	return !wl.DeletionTimestamp.IsZero() ||
+		workload.IsFinished(wl) ||
+		workload.IsEvicted(wl) ||
+		!workload.IsActive(wl) ||
+		!hasSupportedOwner(wl)
+}
+
+func hasSupportedOwner(wl *kueue.Workload) bool {
+	// For now, we only support JobSets.
+	return isJobSetOwner(wl)
+}
+
+func isJobSetOwner(wl *kueue.Workload) bool {
+	if owner := metav1.GetControllerOf(wl); owner != nil {
+		return owner.APIVersion == jobset.SchemeGroupVersion.String() && owner.Kind == "JobSet"
+	}
+	return false
 }
 
-func (r *WorkloadReconciler) isRelevantWorkload(wl *kueue.Workload, log logr.Logger) bool {
+func (r *WorkloadReconciler) cleanupSlice(ctx context.Context, wl *kueue.Workload) (bool, error) {
+	slice := v1alpha1.Slice{}
+	sliceKey := core.SliceKeyFromWorkload(wl)
+
+	log := ctrl.LoggerFrom(ctx).WithValues("slice", klog.KRef(sliceKey.Namespace, sliceKey.Name))
+	ctrl.LoggerInto(ctx, log)
+
+	err := r.client.Get(ctx, sliceKey, &slice)
+	if apierrors.IsNotFound(err) {
+		// slice not found
+		return true, nil
+	} else if err != nil {
+		// error fetching slice
+		log.Error(err, "Failed to fetch the Slice")
+		return false, err
+	}
+
+	terminated, err := r.ownerPodsFinished(ctx, wl)
+	if err != nil || !terminated {
+		return false, err
+	}
+
+	if slice.DeletionTimestamp.IsZero() {
+		return false, client.IgnoreNotFound(r.client.Delete(ctx, &slice))
+	}
+
+	return meta.IsStatusConditionTrue(slice.Status.Conditions, string(v1alpha1.Deformed)), nil
```

Comment:
Can we do this, since we'll be checking for Deformed/Deleted above?
```suggestion
         log("All Pods in workload fininished. Deleting slice")
         if err := client.IgnoreNotFound(r.client.Delete(ctx, &slice)); err != nil && !apierrors.IsNotFound(err) {
           log("Failed to delete slice", err)
           return false, err
         }
         return true, nil
```